### DISCUSSION TITLE
Hard-coded transactions

### DISCUSSION THREAD

**Discussion by jotapea** [CounterpartyXCP/CIPs#137] [General]
Following up on https://github.com/CounterpartyXCP/counterparty-core/issues/1791:

My original question:

>I am reading there is a plan to hard-code dispense transactions, to [remove dependencies](https://github.com/CounterpartyXCP/counterparty-core/issues/1764) of the software.
>
>We already have a hard-coded list of [burns](https://github.com/CounterpartyXCP/counterparty-core/blob/master/counterparty-core/counterpartycore/mainnet_burns.csv).
>
>My question (issue) is, why is this a goal for the current maintainers of the (reference implementation of the) protocol? I can easily understand that hard-coding data can improve the efficiency of some processes... but shouldn't this software be about reading the data written in Bitcoin?
>
>I've been repeated that this can be verified with old versions of the protocol... but isn't the verification itself the PURPOSE of this software? Asking users to run old versions is literal obfuscation, thus REDUCING the trust of the protocol itself.
>
>Counterparty, as a Bitcoin application, should abide by the same principle of "don't trust, verify".

@adamkrellenstein response (and why this is reposted here):

>Hardcoding historical data that can trivially be verified with old versions of the software is not "literal obfuscation", and the trust model is completely unchanged when one is required to do so.
>
>This GitHub repo is not the place for a remedial class in how blockchains work. If you want to learn, try asking for help on the [official forums](https://github.com/CounterpartyXCP/Forum/discussions). If you continue to spam this repo with nonsense, your account will be blocked.

**Comment by jdogresorg** [comment]
>  If you continue to spam this repo with nonsense, your account will be blocked.

WTF?!? So... @jotapea, just to clarify, you went from being added as a core dev in January (after I stepped down), to being removed as a core dev without discussion in March, and now being threatened with a ban from github repo if you continue to post and object too much? 

I guess just letting you [post your opinion](https://github.com/CounterpartyXCP/Forum/discussions/97) as much as you want, and ignoring anything one doesn't want to reply to, is not an option anymore...

Disappointing to consistently see changes being made without any community discussion... API forced to SSL overnight, forums gone, then restored after people object, renaming counterparty-lib breaking fednode installs, CIPs depreciated without any conversation, devs gaslight as being "toxic" for raising legitimate concerns. Unfortunately, this seems to follow the similar narrative that is being pushed in main CP channels, that just a small group of devs are "causing problems and being toxic"... rather than a group of devs, who have put considerable amount of work into building the ecosystem and tools for Counterparty for many years (not talking about myself), objecting to changes without discussions. 

I know this will probably be spun as "devs trying to make trouble" and "devs who dads didn't love them enough"... but this is truly disappointing behavior. This is not how things are supposed to work in a decentralized community project. 

Control of github repo != Control of the ledger

**Reply by jotapea** [reply]
I truly feel being manipulated and used, when convenient.

https://www.counterparty.io/post/counterparty-development-update-january-21-2024

**Reply by jotapea** [reply]
Some don't take criticism very well. They won't even engage and quickly throw out generalizations like "FUD" and "you don't know what you are talking about".

Asking fundamental questions is not allowed. Maybe there is a reason...

Lets hope this "discourse" is allowed to be discussed. Hopefully, focusing on the topic and not "gaslighting" for the non-technical masses.

**Reply by jdogresorg** [reply]
In Adams defense, you can be quite verbose, and it can be very frustrating to continually be questioned again and again on the same issues (you drove me fucking insane with your constant questioning).... but, I believe that is part of the job of the project maintainer, to listen to opposing views and consider them... even if one thinks the viewpoint is off/wrong. 

IMO Ignoring comments is an valid option, personal attacks, gaslighting, blocking/censoring feedback are not.

**Reply by jotapea** [reply]
Fair lol. All part of why we are here now I guess

**Reply by adamkrellenstein** [reply]
@jotopea was never a core dev. When I added him to the repo I mistakenly believed that he had been contributing to it previously.

<img width="457" alt="Screenshot 2024-05-15 at 1 07 25 PM" src="https://github.com/CounterpartyXCP/Forum/assets/6189829/cbbf0278-f5c6-4ae9-b418-4ce380ceb9ef">

The warning was not for objecting, but for posting nonsense to GitHub, which is for substantive discussions on concrete technical matters. 

Also, the API was not forced to SSL overnight. Users and devs were given weeks and weeks of notice. Am happy to provide screenshots. Etc. Etc.

**Reply by jdogresorg** [reply]
![jennifer-lawrence-thumbs-up](https://github.com/CounterpartyXCP/Forum/assets/376028/096eee52-6d15-483c-be71-37c9b14a483e)

**Reply by jdogresorg** [reply]
![image](https://github.com/CounterpartyXCP/Forum/assets/376028/f3ace7ef-22b3-451f-a1bc-ea47f4f9e104)

**Reply by adamkrellenstein** [reply]
<img width="716" alt="Screenshot 2024-05-15 at 2 22 46 PM" src="https://github.com/CounterpartyXCP/Forum/assets/6189829/cd1b7068-f699-4141-b9a7-e8f630f2426e">
<img width="826" alt="Screenshot 2024-05-15 at 2 27 23 PM" src="https://github.com/CounterpartyXCP/Forum/assets/6189829/f427821e-d104-40e7-9144-8c18bd81f277">
<img width="771" alt="Screenshot 2024-05-15 at 2 25 45 PM" src="https://github.com/CounterpartyXCP/Forum/assets/6189829/0626b5ef-d11a-4f53-8d86-abd20b34c774">
<img width="738" alt="Screenshot 2024-05-15 at 2 22 07 PM" src="https://github.com/CounterpartyXCP/Forum/assets/6189829/bac5a8c2-4b6e-4d27-90e6-0d2531ce0349">

I absolutely wish I could have given more notice—or made it backwards-compatible—but the upgrade was rushed because of the network-stopping bug and deployment of v9.x.y was broken.

**Comment by jotapea** [comment]
To follow up on the original topic, my main concern with hardcoded transactions is with the user trust to this metaprotocol. More specifically, from new developers.

In my personal experience, I was let down when I learned about the burns csv. As a Bitcoin user, my expectation is to be able to verify the complete history of Counterparty (Bitcoin) transactions, to learn first-hand about its legitimacy, and build on top of THIS (the legitimacy).

This can be seen, for example, in the client side verification of the xcp.dev user experience.

The argument of "better, faster and more nodes" is deceptive if these are not really fully verifying nodes. Fully verifying this protocol will now REQUIRE the old nodes, in ADDITIONAL effort, resources and time!

The docs specify that the bootstrap should not be used by production nodes, so why is a "hard-coded bootstrap" so acceptable?

"Is not the same as the bootstrap" will be the response to the previous, and is technically not the same. True. BUT in terms of timechain verification, "did the parsing detect everything (and correctly)" case, the csvs are literally a subset of the transactions in the bootstrap. A small subset in the case of the burns. A MUCH larger one in the case of dispenses. BOTH as important IMO, but one with much larger impact (or not? burns, thus XCP, is the root of all named assets...).

What if there is an issue with the ledger?

This is the ONE uncomfortable discussion that should be open. Maybe, some (or most?) think the opposite.

BUT if the sole data source is the immutable timechain, SOMEONE will find the inconsistencies... eventually. Maybe this is "FUD"... but maybe not? Lets verify it!

So, in my view, the "hardcoding strategy" should be about misses in the detection and / or parsing of the transactions (bugs fixed with references to commit history). Instead, the current design of hardcoding, based on what is done in burns csv, is about "trust me, these are the transactions to look at" in Bitcoin. If this is acceptable, then, why not just hardcode the complete list of 2,700,000 transactions? Well, because it is relying on trust on being exhaustive, not on verification.
